airsonic-main/src/main/java/org/airsonic/player/service/sonos/SonosServiceRegistration.java

```diff
@@ -31,75 +30,159 @@
+        APPLICATION_LINK("AppLink");
+
+        private String fieldValue;
+
+        AuthenticationType(String fieldValue){
+            this.fieldValue = fieldValue;
+        }
+
+        public String getFieldValue() {
+            return fieldValue;
+        }
+    }
+
+
+    /**
+     * Enable or disable Sonos registration
+     *
+     * @param airsonicBaseUrl must be the ip address, not the name
+     * @param sonosControllerIp must be a ip too
+     * @param enabled true for enable or false to disable
+     * @param sonosServiceName the name of service you will see on Sonos service list
+     * @param sonosServiceId the ID, the free id is : 240-253 or 255
+     * @throws IOException if some io problem
+     */
+    public boolean setEnabled(String airsonicBaseUrl, String sonosControllerIp, boolean enabled, String sonosServiceName,
+                           int sonosServiceId, AuthenticationType authenticationType) throws IOException {
         String localUrl = airsonicBaseUrl + "ws/Sonos";
         String controllerUrl = String.format("http://%s:1400/customsd", sonosControllerIp);
 
         LOG.info((enabled ? "Enabling" : "Disabling") + " Sonos music service, using Sonos controller IP " + sonosControllerIp +
                  ", SID " + sonosServiceId + ", and Airsonic URL " + localUrl);
 
-        List<Pair<String, String>> params = new ArrayList<Pair<String, String>>();
+        List<Pair<String, String>> params = new ArrayList<>();
         params.add(Pair.create("sid", String.valueOf(sonosServiceId)));
+
+        // Need the csrf token on each request
+        String csrfToken = retrieveCsrfToken(controllerUrl);
+        if(csrfToken != null){
+            params.add(Pair.create("csrfToken", csrfToken));
+        }
+
         if (enabled) {
             params.add(Pair.create("name", sonosServiceName));
             params.add(Pair.create("uri", localUrl));
             params.add(Pair.create("secureUri", localUrl));
             params.add(Pair.create("pollInterval", "1200"));
-            params.add(Pair.create("authType", "UserId"));
             params.add(Pair.create("containerType", "MService"));
             params.add(Pair.create("caps", "search"));
             params.add(Pair.create("caps", "trFavorites"));
             params.add(Pair.create("caps", "alFavorites"));
             params.add(Pair.create("caps", "ucPlaylists"));
             params.add(Pair.create("caps", "extendedMD"));
+
+            // If you change /home/michel/externalDev/airsonic/airsonic-main/src/main/webapp/sonos/presentationMap.xml
+            // Change the presentationMapVersion
             params.add(Pair.create("presentationMapVersion", "1"));
             params.add(Pair.create("presentationMapUri", airsonicBaseUrl + "sonos/presentationMap.xml"));
-            params.add(Pair.create("stringsVersion", "5"));
+
+            // If you are change text in : /home/michel/externalDev/airsonic/airsonic-main/src/main/webapp/sonos/strings.xml
```

Comment:
This is likely wrong :)

Reply:
What is wrong there?

Reply:
If you speek about change the version on string and presentation I can lokking for again in  sonos documentation. Maybe I'm wrong, but when I change the string it's never change on Sonos, if I don't change the version.

Reply:
```suggestion
            // Don't forget to change `stringsVersion` if you change the text in airsonic/airsonic-main/src/main/webapp/sonos/strings.xml
```